fence_cdu: new fence agent

watologo1 opened the pull request:
This fence has been found in a really old Ubuntu fence-agents version
(distro precise from 2012) and was added by a separate tarball.
It has been re-worked to be compatible with latest fence-agents mainline
package, tested against a SENTRY power switch with firmware version 7.1a.

The original code has been written by Andres Rodriguez. Beside adding myself
with a recent copyright notice reflecting my changes, the original license
description on top of the sources (GPL) is untouched.

No idea why this never made it mainline, but I can confirm it convenient
and working with the HW I tested against.

Main credits for the original fence development go to:

CC: Andres Rodriguez <andres.rodriguez@canonical.com>

oalbrigt commented on agents/cdu/fence_cdu.py:
```diff
@@ -0,0 +1,143 @@
+            fail(EC_TIMED_OUT)
+
+def main():
+    device_opt = [  "help", "version", "agent", "quiet", "verbose", "debug",
```
You should be able to remove lots of these (some are included by default, and others might be included by e.g. secure or similar (like inetX_only).

You can run `make xml-check` to see if you removed too many :)

watologo1 replied:
I removed most of it, including all ssh related options.
the old default -1 blowfish option showed that this won't work nowadays anymore anyway and it was mentioned untested at these times already.

oalbrigt commented on agents/cdu/fence_cdu.py:
```diff
@@ -0,0 +1,143 @@
+    ## connection will be closed before we run close(). This is not
+    ## a problem because everything is checked before.
+    ######
+    try:
```
You should move this to a function, and add it as an exit handler:
https://github.com/ClusterLabs/fence-agents/blob/8d746be92f191aa289f13a3703031c122a5e6cf3/agents/vmware_rest/fence_vmware_rest.py#L222

oalbrigt commented on fence-agents.spec.in:
```diff
@@ -41,6 +41,7 @@ fence-agents-aws \\
 fence-agents-azure-arm \\
 fence-agents-bladecenter \\
 fence-agents-brocade \\
+fence-agents-cdu \\
```
You have to add %package/%description/%files sections for it as well.

oalbrigt commented on agents/cdu/fence_cdu.py:
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/python3
```
This needs to be `#!@PYTHON@ -tt` to get set based on detected or specified value during configure/build.

oalbrigt commented on agents/cdu/fence_cdu.py:
```diff
@@ -0,0 +1,143 @@
+#####
+
+import sys, re, pexpect, atexit
+sys.path.append("/usr/share/fence")
```
This also needs to be changed to `@FENCEAGENTSLIBDIR@` to get replaced with detected/specified path when you build the project.

oalbrigt commented on agents/cdu/fence_cdu.py:
```diff
@@ -0,0 +1,137 @@
+    conn.close()
+
+def main():
+    device_opt = [  "action", "ipaddr", "login", "secure", "port",
```
You can also remove action and ipport here.

watologo1 commented:
Should be better now... ?
Thanks for your quick review(s)!

oalbrigt commented:
ok to test

oalbrigt commented:
Looking good now. Just a bit too much cleanup, so just readd "secure" in device_opt and run `make xml-check` to see that it only complains about difference in description (and then run `make xml-upload`.

watologo1 commented:
Next round:
Added secure, new xml test file, due to previous description modifs (removed ssh comments) and also added manpage to the file section this time.

oalbrigt commented:
Nice.

I'll do some testing to find what needs changing to pass that delay-check :)

oalbrigt commented:
You need to add "passwd" to device_opt (so you'll have to run `make xml-upload` again after doing so).

You can also add the "no_login", "no_password" combination if the devices support using SSH keys to login.

watologo1 commented:
Am Freitag, 25. Juni 2021, 09:37:04 CEST schrieb Oyvind Albrigtsen:
> You need to add "passwd" to device_opt (so you'll have to run `make
> xml-upload` again after doing so).

Ah, yes. No password option anymore.
 
> You can also add the "no_login", "no_password" combination if the devices
> support using SSH keys to login.

I realized that the "secure" option implies:
--ssh, --identify-file and some more.

I tried a bit, but I could not get ssh connection working (also not a native
one using ssh executable).

First I run into:
no matching key exchange method found. Their offer: diffie-hellman-group1-sha1

then after adding:
-oKexAlgorithms=+diffie-hellman-group1-sha1
I run into:
no matching host key type found. Their offer: ssh-dss

I went the Sentry configs up and down, but I could not find anything related 
to ssh keys, neither in the ssh, nor in the user or whatever options.

So I would leave the "secure" option out.
It can be added later if I or someone else finds out more...

Sending another version with "passwd" added...

Thanks,

  Thomas

oalbrigt commented:
Sounds good to me.

oalbrigt commented:
LGTM. Thanks.

watologo1 commented:
> LGTM. Thanks.

Thank you. This was a nice and efficient review/work. It's very much appreciated!
